### PR TITLE
Fix #10573: Devirtualize member selection when matching

### LINKED ISSUE #10573: Quoted pattern should devirtualize member selections
The semantics still seem incomplete in this implementation.

Consider:

```scala
trait A { def foo: Int }
trait B extends A { def foo: Int = 1 }

val e0: Expr[A] = '{ new B }
val e1: Expr[Int] = '{ $e0.foo }
```

The quoted expression `e1` will not match a pattern like `'{ ($x: B).foo }`, but it should.

_Originally posted by @LPTK in https://github.com/lampepfl/dotty/pull/10531#discussion_r532775962_

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance with #quotes please

File: compiler/src/scala/quoted/runtime/impl/Matcher.scala [hunk 1]
```diff
@@ -348,10 +348,21 @@ object Matcher {
...
+          qual.tpe.typeSymbol.methods.find( sym =>
+            sym.name == scrutinee.name &&
+            sym.allOverriddenSymbols.contains(scrutinee)
+          ).getOrElse(scrutinee)
+        case _ => scrutinee
+      val pattern = patternTree.symbol
+
+      devirtualizedScrutinee == pattern
+      || summon[Env].get(devirtualizedScrutinee).contains(pattern)
+      || devirtualizedScrutinee.allOverriddenSymbols.contains(pattern)
```
Comment (liufengyun) [inline compiler/src/scala/quoted/runtime/impl/Matcher.scala:365]: The document needs to be updated. The following code might be faster:

```Scala
scrutinee.overridingSymbol(pattern.owner) == pattern
|| pattern.overridingSymbol(scrutinee.owner) == scrutinee
```

**Review by liufengyun — APPROVED** [review]
Otherwise, LGTM

File: compiler/src/scala/quoted/runtime/impl/Matcher.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/Matcher.scala:365]: Will try that. The current performance is unaccepable.

**Comment by nicolasstucki** [comment]
test performance with #quotes please

File: compiler/src/scala/quoted/runtime/impl/Matcher.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/Matcher.scala:363]: `overridingSymbol` worked, but not there.

Comment (liufengyun) [reply compiler/src/scala/quoted/runtime/impl/Matcher.scala:363]: If the check is whether the two symbols override each other, can we simply replace the body of the method with the following:

```Scala
scrutinee.overridingSymbol(pattern.owner) == pattern
|| pattern.overridingSymbol(scrutinee.owner) == scrutinee
```

Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/Matcher.scala:363]: That version had some false positives involving modules.

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
@liufengyun the benchmarks might be stuck

**Comment by liufengyun** [comment]
Yes, I stopped it, I am investigating the latest failures.

**Comment by nicolasstucki** [comment]
test performance with #quotes please

**Comment by nicolasstucki** [comment]
Now performance is fine

**Review by liufengyun — APPROVED** [review]
The document for `symbolMatch` needs to be updated. Otherwise, LGTM

**Pull request merged**